### PR TITLE
Complete Auth

### PR DESCRIPTION
**Pull request by SADmammoth**
Date of last feature commit: 13.04.2020
#### Implemented:
* LoginForm
* New member page
* Authorisation and authentication
* User roles context
* Rebuild components according roles
* Header navigation

### CODE REVIEW

File: src/helpers/AuthenticationManager.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    super(props);
+    this.state = { authenticated: null };
+  }
+
+  async componentDidMount() {
+    await this.logIn();
+  }
+
+  logIn = async (login, password) => {
+    this.setState({ authenticated: (await this.authenticate(login, password)).status === 'success' });
```
Comment (maximsan) [inline src/helpers/AuthenticationManager.js:17]: split in 2 acts

File: src/helpers/AuthenticationManager.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+      }
+
+      localStorage.setItem('authToken', token);
+      this.props.authorize(role, userID);
+    } else {
+      const { role, userID } = await Client.getUserInfoByToken(authToken);
+      this.props.authorize(role, userID);
+    }
+    return { status: 'success', message: 'Login successful' };
+  };
```
Comment (maximsan) [inline src/helpers/AuthenticationManager.js:41]: authenticate can be helper method

File: src/helpers/FirebaseFaker.js [hunk 3]
```diff
@@ -36,29 +37,29 @@ export default class FirebaseFaker {
     members.forEach((el) => Client.db.collection('memberTasks').add(el));
   }
 
-  static generateProgress(userID) {
-    let members = new Array(5).fill(null);
+  // static generateProgress(userID) {
+  //   let members = new Array(5).fill(null);
```
Comment (maximsan) [inline src/helpers/FirebaseFaker.js:41]: remove ?

File: src/helpers/calculateAge.js [hunk 4]
```diff
@@ -1,3 +1,12 @@
 export default function calculateAge(birthDate) {
+  let now = new Date();
+  if (now.getMonth() === birthDate.getMonth()) {
+    if (now.getDate() > birthDate.getDate()) {
+      return new Date().getFullYear() - birthDate.getFullYear() - 1;
+    }
+  }
+  if (now.getMonth() < birthDate.getMonth()) {
+    return new Date().getFullYear() - birthDate.getFullYear() - 1;
```
Comment (maximsan) [inline src/helpers/calculateAge.js:9]: don't repeat new Date().getFullYear() - birthDate.getFullYear() - 1

File: src/pages/MemberTracksPage.js [hunk 5]
```diff
@@ -50,9 +52,13 @@ class MemberTracksPage extends React.Component {
...
-        <Header>
-          <h1>Your Task tracks</h1>
-        </Header>
+        <UserContext>
+          {({ role, userID }) => {
+            return (
+              <Header role={role} title='Task tracks' navItems={getNavItems({ role, userID }, this.props.match.path)} />
+            );
+          }}
+        </UserContext>
```
Comment (maximsan) [inline src/pages/MemberTracksPage.js:61]: create HOC withUserContext

File: src/pages/MembersManagerPage.js [hunk 6]
```diff
@@ -31,52 +33,73 @@ class MembersManagerPage extends React.Component {
...
+          );
+        }}
+      </UserContext.Consumer>
     );
   }
 
+  static renderMember(id, data) {
+    const WrappedMemberCard = MembersManagerPage.WrappedMemberCard;
+    return <WrappedMemberCard id={id} {...data} />;
+  }
```
Comment (maximsan) [inline src/pages/MembersManagerPage.js:88]: why do you need static method ?

File: src/components/elements/Form/InputMask.js [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+import Validator from '../../../helpers/Validator';
+
+function MaskedInput(input, mask, validate = false, type = 'default') {
+  let resultInput = input;
+  console.log(validate);
+  console.log(input.props.value);
+  if (input.props.type === 'textarea' || input.props.type === 'select') {
+    throw Error('TextArea and SelectInput do not support input mask');
+  }
+  const maskArray = mask.split(/(a\\(?!\\))|(h\\(?!\\))|(#\\(?!\\))|(9\\(?!\\))|(\\\\)|()/).filter((el) => el);
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:12]: const variable

File: src/components/elements/Form/InputMask.js [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+    default:
+      return followMask(resultInput, maskArray);
+  }
+}
+
+function followInvisibleMaskComponent(input, maskArray) {
+  return React.cloneElement(input, {
+    onFocus: (event) => {
+      if (!event.target.value || event.target.value === '') {
+        const firstPlaceholder = maskArray.findIndex((el) => /[9aAh%#](?!(\\)(?!\\))/.test(el));
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:34]: const variable

File: src/components/elements/Form/InputMask.js [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+        event.target.value = maskArray
+          .splice(0, firstPlaceholder)
+          .join('')
+          .replace(/([9aAh%#])\\(?!\\)/s, '$1');
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:38]: const variable

File: src/components/elements/Form/InputMask.js [hunk 10]
```diff
@@ -0,0 +1,182 @@
...
+        event.target.value = value.join('');
+        event.preventDefault();
+      }
+    },
+  });
+}
+
+function followInvisibleMask(input, maskArray) {
+  if (maskArray[input.props.value.length]) {
+    if ((input.props.value && input.props.value !== '') || /[9aAh%#](?!(\\)(?!\\))/.test(maskArray[0])) {
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:65]: const variable

File: src/components/elements/Form/InputMask.js [hunk 11]
```diff
@@ -0,0 +1,182 @@
...
+      if (!/[9aAh%#](?!(\\)(?!\\))/.test(maskArray[input.props.value.length])) {
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:66]: ****

Comment (maximsan) [inline src/components/elements/Form/InputMask.js:66]: const variable

File: src/components/elements/Form/InputMask.js [hunk 12]
```diff
@@ -0,0 +1,182 @@
...
+        value.splice(
+          start,
+          end - start,
+          target.value.substring(start, end).replace(
+            new RegExp(
+              `[^${maskArray
+                .filter((el, i, arr) => !arr.slice(i + 1).includes(el))
+                .join('')
+                .replace(/[9aAh%#](?!(\\)(?!\\))/g, '')
+                .replace(/([9aAh%#])\\(?!\\)/, '$1')}]+`,
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:113]: const variable

File: src/components/elements/Form/InputMask.js [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+        value = value.split('');
+        value.splice(
+          start,
+          end - start,
+          target.value.substring(start, end).replace(
+            new RegExp(
+              `[^${maskArray
+                .filter((el, i, arr) => !arr.slice(i + 1).includes(el))
+                .join('')
+                .replace(/[9aAh%#](?!(\\)(?!\\))/g, '')
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:112]: const variable

File: src/components/elements/Form/InputMask.js [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+      return changedInput;
+    }
+    // if (!/[9aAh%#](?!(\\)(?!\\))/.test(maskArray[input.props.value.length])) {
+    //   changedInput.props.onInput({
+    //     target: {
+    //       name: input.props.name,
+    //       value: input.props.value + maskArray[input.props.value.length].replace(/([9aAh%#])\\(?!\\)/g, '$1'),
+    //     },
+    //   });
+    // }
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:164]: remove

File: src/components/elements/Form/InputMask.js [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+  return string + mask.replace(/[9aAh%#](?!(\\)(?!\\))/g, '_').replace(/([9aAh%#])\\(?!\\)/, '$1');
+}
+// MaskedInput.defaultProps = {
+//   type: 'default',
+// };
+
+// MaskedInput.propTypes = {
+//   type: PropTypes.oneOf(['invisible', 'default']),
+//   children:
+// };
```
Comment (maximsan) [inline src/components/elements/Form/InputMask.js:180]: remove

File: src/helpers/Validator.js [hunk 16]
```diff
@@ -26,8 +26,6 @@ export default class Validator {
 
   //
 
-  //
-
   static passwordMessage =
```
Comment (maximsan) [inline src/helpers/Validator.js:29]: no need make it static

File: src/helpers/Validator.js [hunk 17]
```diff
@@ -98,6 +96,8 @@ export default class Validator {
 
   //
 
+  //
+
   static dateTimeMessage = 'Invalid date format';
```
Comment (maximsan) [inline src/helpers/Validator.js:101]: no need make it static

File: src/tests/helpers/addMask.test.js [hunk 18]
```diff
@@ -0,0 +1,17 @@
+import maskEscapedCharsOrEmptyRegex from '../../helpers/maskEscapedCharsOrEmptyRegex';
+import addMask from '../../helpers/addMask';
+
+describe('addMask', () => {
+  test('Adds mask correct', () => {
+    const maskArray = '+375 (9\\9\\) 999-99-99'.split(maskEscapedCharsOrEmptyRegex).filter((el) => el);
+    const input = '+375 (99) 2';
```
Comment (maximsan) [inline src/tests/helpers/addMask.test.js:7]: add space before expected everywhere

File: src/tests/helpers/addMask.test.js [hunk 19]
```diff
@@ -0,0 +1,17 @@
...
+    const expectedOutput = '+375 (99) 2__-__-__';
```
Comment (maximsan) [inline src/tests/helpers/addMask.test.js:8]: add space before expect everywhere

File: src/tests/helpers/calculateAge.test.js [hunk 20]
```diff
@@ -0,0 +1,21 @@
+import calculateAge from '../../helpers/calculateAge';
+
+describe('calculateAge', () => {
+  test('Calculates age correct, if cake day is before current date', () => {
+    const now = new Date();
+    const birthDate = new Date(2000, 7, 18);
+    expect(calculateAge(birthDate)).toBe(now.getFullYear() - 2000 - 1);
```
Comment (maximsan) [inline src/tests/helpers/calculateAge.test.js:7]: now.getFullYear() - 2000 - 1 should be saved in expected variable
calculateAge(birthDate) in actual

File: src/tests/helpers/calculateAge.test.js [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+  });
+
+  test('Calculates age correct, if cake day is on current date', () => {
```
Comment (maximsan) [inline src/tests/helpers/calculateAge.test.js:10]: if cake day is on current date - ???

File: src/tests/helpers/checkboxValueSeparator.test.js [hunk 22]
```diff
@@ -0,0 +1,16 @@
+import checkboxValueSeparator from '../../helpers/checkboxValueSeparator';
+
+describe('checkboxValueSeparator', () => {
+  test('Returns correct array for string input', () => {
+    expect(checkboxValueSeparator('value1,value2,value3,value4')).toStrictEqual([
+      'value1',
+      'value2',
+      'value3',
+      'value4',
+    ]);
```
Comment (maximsan) [inline src/tests/helpers/checkboxValueSeparator.test.js:10]: split in actual, expected and assert

File: src/tests/helpers/checkboxValueSeparator.test.js [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+  });
+  test("Returns input if it's array", () => {
+    const inputArray = ['value1', 'value2', 'value3', 'value4'];
+    expect(checkboxValueSeparator(inputArray)).toBe(inputArray);
```
Comment (maximsan) [inline src/tests/helpers/checkboxValueSeparator.test.js:14]: split in actual, expected and assert

File: src/tests/helpers/compareObjects.test.js [hunk 24]
```diff
@@ -0,0 +1,20 @@
+import compareObjects from '../../helpers/compareObjects';
+
+describe('compareObjects', () => {
+  test('Returns true if references are same', () => {
+    const object1 = { key1: 'value', key2: 'value2', key3: 'value3' };
```
Comment (maximsan) [inline src/tests/helpers/compareObjects.test.js:5]: create deeper object

File: src/tests/helpers/invisibleMaskOnInputValue.test.js [hunk 25]
```diff
@@ -0,0 +1,14 @@
...
+describe('invisibleMaskOnInputValue', () => {
+  test('Return value is object with correct structure', () => {
+    const result = invisibleMaskOnInputValue('name', 'value', ['v', 'a', 'l', 'u', 'e', 's']);
+    expect(result).toHaveProperty('target');
+    expect(result.target).toHaveProperty('name');
+    expect(result.target).toHaveProperty('value');
+  });
+  test('Return value must add not placeholder characters to input', () => {
+    const result = invisibleMaskOnInputValue('name', 'value', ['v', 'a', 'l', 'u', 'e', 's']);
+    expect(result.target.value).toBe('values');
```
Comment (maximsan) [inline src/tests/helpers/invisibleMaskOnInputValue.test.js:12]: split in actual, expected and assert

File: src/tests/helpers/invisibleMaskOnInputValue.test.js [hunk 26]
```diff
@@ -0,0 +1,14 @@
+import invisibleMaskOnInputValue from '../../helpers/invisibleMaskOnInputValue';
+
+describe('invisibleMaskOnInputValue', () => {
+  test('Return value is object with correct structure', () => {
+    const result = invisibleMaskOnInputValue('name', 'value', ['v', 'a', 'l', 'u', 'e', 's']);
+    expect(result).toHaveProperty('target');
+    expect(result.target).toHaveProperty('name');
+    expect(result.target).toHaveProperty('value');
```
Comment (maximsan) [inline src/tests/helpers/invisibleMaskOnInputValue.test.js:8]: split in actual, expected and assert

File: src/tests/helpers/maskEscapedCharsRegex.test.js [hunk 27]
```diff
@@ -0,0 +1,7 @@
+import maskEscapedCharsRegex from '../../helpers/maskEscapedCharsRegex';
+
+describe('maskEscapedCharsRegex', () => {
+  test('Must match escaped chars', () => {
+    expect('9\\9\\\\8312a\\'.match(maskEscapedCharsRegex)).toStrictEqual(['9\\', '\\\\', 'a\\']);
```
Comment (maximsan) [inline src/tests/helpers/maskEscapedCharsRegex.test.js:5]: split in actual, expected and assert

File: src/tests/helpers/maskSpecialCharsRegex.test.js [hunk 28]
```diff
@@ -0,0 +1,7 @@
+import maskSpecialCharsRegex from '../../helpers/maskSpecialCharsRegex';
+
+describe('maskEscapedCharsRegex', () => {
+  test('Must match special chars and not include escaped', () => {
+    expect('9\\9\\\\831aa\\'.match(maskSpecialCharsRegex)).toStrictEqual(['9', 'a']);
```
Comment (maximsan) [inline src/tests/helpers/maskSpecialCharsRegex.test.js:5]: split in actual, expected and assert

File: src/components/elements/LoginForm/LoginForm.js [hunk 29]
```diff
@@ -0,0 +1,25 @@
...
+const LoginForm = (props) => {
+  return (
+    <>
+      <div className='login'>
+        <div className='content'>
+          <h1>Welcome to DIMSUI</h1>
+          <Form
+            className='login-form'
+            inputs={loginFormInputAttributes()}
+            onSubmit={(data) => props.logIn(data.userName, data.password)}
```
Comment (maximsan) [inline src/components/elements/LoginForm/LoginForm.js:16]: create method

File: old.js [hunk 30]
```diff
@@ -0,0 +1,86 @@
...
+(?=.*[A-Z].*)\
+${/* Contain at least 1 uppercase letter */ ''}\
+(?=.*[a-z].*)\
+${/* Contain at least 1 lowercase letter */ ''}\
+(?=.*_.*)\
+${/* Contain at least 1 underscore sign */ ''}\
+(?=.*[0-9].*)\
+${/* Contain at least 1 number */ ''}\
+)(.*)\
+$)`;
```
[the remark this answers is not in the document]
Comment (SADmammoth) [reply old.js:86]: Sorry, old.js was in .gitignore

File: old.js [hunk 31]
```diff
@@ -0,0 +1,86 @@
...
+
+  static phone = (phone, masks = ['#* ###-##-##']) => {
+    Validator.phoneLastMasks = masks;
+    console.log(
+      RegExp(
+        `(^${masks
+          .map((mask) =>
+            RegexpUnescape(mask)
+              .replace(/#\\\*/g, '[0-9]{2,4}')
+              .replace(/#/g, '[0-9]'),
```
[the remark this answers is not in the document]
Comment (SADmammoth) [reply old.js:21]: Sorry, old.js was in .gitignore

**Pull request merged**